bestie/server/unzipper.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// Source:  https://gosamples.dev/unzip-file/  (slightly modified)
+
+// Unzip the contents of the source file to the destination location.
+// Note that any subdirectories contained within the zip file are
+// created within the destination location as a subtree.
+func unzipSource(source, destination string) ([]string, error) {
+	reader, err := zip.OpenReader(source)
+	if err != nil {
+		return nil, err
+	}
+	defer reader.Close()
+
+	// Get the absolute path of the destination.
+	destination, err = filepath.Abs(destination)
+	if err != nil {
+		return nil, err
+	}
+
+	// Process each element contained within the zip file. These can be
+	// a mix of directories and files.
+	var files []string
+	for _, f := range reader.File {
+		err := unzipFile(f, destination)
+		if err != nil {
+			return nil, err
+		}
+		if !f.FileInfo().IsDir() {
+			files = append(files, filepath.Join(destination, f.Name))
+		}
+	}
+
+	return files, nil
+}
+
+func unzipFile(f *zip.File, destination string) error {
+	// Guard against "zip slip" vulnerability.
+	filePath := filepath.Join(destination, f.Name)
+	if !strings.HasPrefix(filePath, filepath.Clean(destination)+string(os.PathSeparator)) {
+		return fmt.Errorf("invalid file path: %s", filePath)
+	}
+
+	// If element is a directory, create it in the destination tree and return.
+	if f.FileInfo().IsDir() {
+		if err := os.MkdirAll(filePath, os.ModePerm); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// For a file element, make sure its path exists.
+	if err := os.MkdirAll(filepath.Dir(filePath), os.ModePerm); err != nil {
+		return err
+	}
```

Comment:
I suspect the biggest improvement that can be made to this code is to not actually unzip the file via Go - instead, we can iterate over the zip to find interesting files, and selectively either read those files into memory, or spill them to temporary files to read. This would help by:
* not having to write/maintain essentially a fork of the zip tool in Go
* not having to lock/unlock in the service (since the service won't contend on the same destination paths)
* not having to worry about overly large zips
* not having to worry about cleaning up zips

Maybe this can be replaced with a function that:
* iterates through the zip file headers
* if the current zip file header is an interesting file (metrics.pb) - do something with it
  * maybe read the file contents into a buffer, and then unmarshal the pb message

Reply:
@gpaussaenf a toy example of this is https://gist.github.com/indraniel/1a91458984179ab4cf80

Reply:
Thanks for the link. Looks like `tarReader.Read(data)` directly into a []byte slice might be what I need.

Reply:
Made this change in the latest commit.